src/soar_sdk/views/view_parser.py

```diff
@@ -0,0 +1,123 @@
+from typing import (
+    Any,
+    Callable,
+    TypeVar,
+    Union,
+    get_origin,
+    get_args,
+    Generic,
+    cast,
+)
+import contextlib
+import inspect
+from soar_sdk.action_results import ActionOutput, ActionResult
+from soar_sdk.models.view import ViewContext, AppRunMetadata
+
+T = TypeVar("T", bound=ActionOutput)
+
+
+AllAppRuns = list[tuple[AppRunMetadata, list[ActionResult]]]
+
+
+class ViewFunctionParser(Generic[T]):
+    """Handles parsing and validation of view function signatures and execution."""
+
+    def __init__(
+        self,
+        function: Callable,
+    ) -> None:
+        self.function = function
+
+        # Auto-detect output class from function signature
+        detected_class = self.auto_detect_output_class(function)
+        self.output_class: type[T] = cast(type[T], detected_class)
+
+    @staticmethod
+    def auto_detect_output_class(function: Callable) -> type[ActionOutput]:
+        """Auto-detect ActionOutput class from function type annotations."""
+        signature = inspect.signature(function)
+
+        for param in signature.parameters.values():
+            if param.annotation == inspect.Parameter.empty:
+                continue
+
+            origin = get_origin(param.annotation)
+            if origin is list:
```

Comment:
Nit: make it more obvious that we don't use `origin` again:
```suggestion
            if get_origin(param.annotation) is list:
```

Reply:
Updated